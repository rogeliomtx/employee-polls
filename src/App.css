/* Auth */
body {
  background-color: #f5f5f5;
}

.card.auth {
  width: 18rem;
}

img.avatar, div.avatar {
  width: 100px;
  height: 100px;
  display: block;
  margin: 0 auto;
  border-radius: 100%;
}

div.avatar {
  background: #0d6efd;
  color: #fff;
  border-radius: 100%;
}

div.avatar span {
  font-size: 3em;
  line-height: 100px;
  display: block;
  text-align: center;
}

.navbar .avatar {
  width: 30px;
  height: 30px;
  display: inline-block;
  margin-right: .5em;
}

.leaderboard .avatar {
  width: 50px;
  height: 50px;
  display: inline-block;
  margin-right: .5em;
}

.navbar .logout {
  cursor: pointer;
}