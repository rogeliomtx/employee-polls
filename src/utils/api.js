import { _getUsers, _getQuestions, _saveQuestion, _saveQuestionAnswer } from './_DATA';

export async function getInitialData () {
  return Promise.all([
    _getUsers(),
    _getQuestions(),
  ]).then(([users, questions]) => ({
    users,
    questions
  }))
}

export async function saveQuestion (question) {
  return _saveQuestion(question)
}

export async function saveQuestionAnswer ({ authedUser, question, answer }) {
  return _saveQuestionAnswer({
    authedUser,
    qid: question.id,
    answer
  }).then((saved) => {
    return { authedUser, answer, questionId: question.id }
  })
}

export async function login (id, password) {
  return _getUsers().then((users) => {
    const user = Object.values(users).find((user) => user.id === id && user.password === password);
    return user;
  })
}
