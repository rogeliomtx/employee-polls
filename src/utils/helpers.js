export function formatQuestion (question, author, authedUser) {
  const { id, optionOne, optionTwo, timestamp } = question

  return {
    id,
    timestamp,
    optionOne,
    optionTwo,
    author,
    hasAnswered: optionOne.votes.includes(authedUser) || optionTwo.votes.includes(authedUser)
  }
}