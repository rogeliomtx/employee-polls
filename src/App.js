import { Route, Routes } from "react-router-dom";
import { useEffect, Fragment } from "react";
import { connect } from "react-redux";
import { LoadingBar } from "react-redux-loading-bar";

import "bootstrap/dist/css/bootstrap.min.css";
import './App.css';

import Home from './components/Home';
import Leaderboard from './components/Leaderboard';
import Login from './components/Login';
import CreateQuestion from './components/CreateQuestion';
import QuestionDetails from './components/QuestionDetails';
import Navbar from './components/Navbar';
import NotFound from './components/NotFound';
import ProtectedRoute from "./components/ProtectedRoute";

import { handleInitialData } from "./actions/shared";

function App({ dispatch, loading }) {
  useEffect(() => {
    dispatch(handleInitialData());
  });

  return (
    <Fragment>
      <LoadingBar />
      {
        loading ? <div>Loading...</div> : (
          <div>
            <Navbar />
            <div className="container">
              <Routes>                
                <Route path="/" element={
                  <ProtectedRoute>
                    <Home />
                  </ProtectedRoute>
                } />
                <Route path="/login" element={<Login />} />
                <Route path="/leaderboard" element={
                  <ProtectedRoute>
                    <Leaderboard />
                  </ProtectedRoute>
                } />
                <Route path="/add" element={
                  <ProtectedRoute>
                    <CreateQuestion />
                  </ProtectedRoute>
                } />
                <Route path="/questions/:questionId" element={
                  <ProtectedRoute>
                    <QuestionDetails />
                  </ProtectedRoute>
                } />
                <Route path="/*" element={<NotFound />} />
              </Routes>
            </div>
          </div>
        )
      }
    </Fragment>
  );
}

const mapStateToProps = ({ users, authedUser }) => {
  return {
    loading: users === null,  // if users are loaded, loading is false
    authedUser,
  }
}
export default connect(mapStateToProps)(App);
