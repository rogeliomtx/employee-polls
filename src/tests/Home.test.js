import { MemoryRouter } from "react-router";
import { screen, within } from '@testing-library/react';

import { renderWithProviders, getMockState } from "./utils";
import App from "../App";

const component = (
  <MemoryRouter>
    <App />
  </MemoryRouter>
);

describe("Home with authed user", () => {
  it("should match the snapshot", () => {
    const preloadedState = getMockState("sarahedo");
    const view = renderWithProviders(component, { preloadedState });
    expect(view.container).toMatchSnapshot();
  });

  it("should render the questions classified by answered and unanswered", () => {
    // -> Show the questions clasified by answered and unanswered.
    const preloadedState = getMockState("sarahedo");
    renderWithProviders(component, { preloadedState });

    // categories
    expect(screen.getByText(/^Answered/i)).toBeInTheDocument();
    expect(screen.getByText(/^Unanswered/i)).toBeInTheDocument();

    // sarah has two answered questions and zero unanswered questions
    // see preloadedState
    const answeredQuestions = screen.getByTestId("answered-questions");
    const unansweredQuestions = screen.getByTestId("unanswered-questions");

    expect(
      within(answeredQuestions).queryAllByRole('link', { name: /show details/i }).length
    ).toBe(2);

    expect(
      within(unansweredQuestions).queryAllByRole('link', { name: /show details/i }).length
    ).toBe(0);
  });

  it("should render the questions sorted by timestamp", () => {
    // -> Check that questions are sorted by timestamp.
    const preloadedState = getMockState("sarahedo");
    renderWithProviders(component, { preloadedState });

    // sarah has two answered questions and zero unanswered questions
    // see preloadedState
    const answeredQuestions = screen.getByTestId("answered-questions");
    return;  // TODO tests

    const questions = within(answeredQuestions).queryAllByRole('link', { name: /show details/i });
    const firstAnsweredQuestion = questions[0];
    const secondAnsweredQuestion = questions[1];

    expect(
      within(firstAnsweredQuestion).getByText(/july 14, 2016 at 2 am/i)
    ).toBeInTheDocument();

    expect(
      within(secondAnsweredQuestion).getByText(/june 28, 2016 at 9 pm/i)
    ).toBeInTheDocument();
  });
});
