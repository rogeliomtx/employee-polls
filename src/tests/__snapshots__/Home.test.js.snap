// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home with authed user should match the snapshot 1`] = `
<div>
  <div />
  <div>
    <nav
      class="navbar navbar-expand-lg bg-primary"
      data-testid="menu"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand link-light"
          href="/"
        >
          Employee Polls
        </a>
        <div
          class="collapse navbar-collapse"
          id="navbarText"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link active link-light"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/add"
              >
                Add question
              </a>
            </li>
          </ul>
          <ul
            class="navbar-nav mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="Avatar"
                    class="avatar"
                    src="/assets/avatar_1.jpg"
                  />
                </div>
                Sarah Edo
              </span>
            </li>
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light logout"
              >
                Logout
              </span>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        role="contentinfo"
      >
        <div
          class="card mt-5"
        >
          <div
            class="card-header"
          >
            Unanswered
          </div>
          <div
            class="card-body"
          >
            <div
              class="row row-cols-4"
              data-testid="unanswered-questions"
            />
          </div>
        </div>
        <div
          class="card mt-5 mb-5"
        >
          <div
            class="card-header"
          >
            Answered
          </div>
          <div
            class="card-body"
          >
            <div
              class="row row-cols-4"
              data-testid="answered-questions"
            >
              <div
                class="column"
              >
                <div
                  class="card text-center mb-3"
                  role="article"
                >
                  <div
                    class="card-body"
                  >
                    <div
                      class="avatar"
                    >
                      <img
                        alt="Avatar"
                        class="avatar"
                        src="/assets/avatar_1.jpg"
                      />
                    </div>
                    <h5
                      class="card-title mb-0"
                    >
                      Sarah Edo
                    </h5>
                    <p
                      class="card-text text-secondary"
                    >
                      July 14, 2016 at 2 AM
                    </p>
                    <a
                      class="btn btn-outline-primary"
                      href="/questions/6ni6ok3ym7mf1p33lnez"
                    >
                      Show details
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="column"
              >
                <div
                  class="card text-center mb-3"
                  role="article"
                >
                  <div
                    class="card-body"
                  >
                    <div
                      class="avatar"
                    >
                      <img
                        alt="Avatar"
                        class="avatar"
                        src="/assets/avatar_1.jpg"
                      />
                    </div>
                    <h5
                      class="card-title mb-0"
                    >
                      Sarah Edo
                    </h5>
                    <p
                      class="card-text text-secondary"
                    >
                      June 28, 2016 at 9 PM
                    </p>
                    <a
                      class="btn btn-outline-primary"
                      href="/questions/8xf0y6ziyjabvozdd253nd"
                    >
                      Show details
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
